File: js/sdk/image.js
import { RAW_FORMAT } from "../core/protocol.js";

function toU8(data) {
  if (data instanceof Uint8Array) return data;
  if (data instanceof ArrayBuffer) return new Uint8Array(data);
  if (ArrayBuffer.isView(data)) return new Uint8Array(data.buffer, data.byteOffset, data.byteLength);
  return new Uint8Array(data || []);
}

export function decodeRawToRgb(frame, rawFormat = RAW_FORMAT) {
  const width = Number(frame?.width || 0);
  const height = Number(frame?.height || 0);
  const data = frame?.data instanceof Uint8Array ? frame.data : null;
  if (!width || !height || !data) return null;

  const out = new Uint8ClampedArray(width * height * 4);
  const format = Number(frame?.format || 0);

  if (format === rawFormat.GRAY8) {
    const need = width * height;
    if (data.length < need) return null;
    for (let i = 0; i < need; i += 1) {
      const v = data[i];
      const j = i * 4;
      out[j] = v;
      out[j + 1] = v;
      out[j + 2] = v;
      out[j + 3] = 255;
    }
    return { width, height, rgba: out };
  }

  if (format === rawFormat.RGB24) {
    const need = width * height * 3;
    if (data.length < need) return null;
    for (let i = 0, j = 0; i < need; i += 3, j += 4) {
      out[j] = data[i];
      out[j + 1] = data[i + 1];
      out[j + 2] = data[i + 2];
      out[j + 3] = 255;
    }
    return { width, height, rgba: out };
  }

  if (format === rawFormat.BGR24) {
    const need = width * height * 3;
    if (data.length < need) return null;
    for (let i = 0, j = 0; i < need; i += 3, j += 4) {
      out[j] = data[i + 2];
      out[j + 1] = data[i + 1];
      out[j + 2] = data[i];
      out[j + 3] = 255;
    }
    return { width, height, rgba: out };
  }

  if (format === rawFormat.RGBA32) {
    const need = width * height * 4;
    if (data.length < need) return null;
    out.set(data.slice(0, need));
    return { width, height, rgba: out };
  }

  if (format === rawFormat.BGRA32) {
    const need = width * height * 4;
    if (data.length < need) return null;
    for (let i = 0; i < need; i += 4) {
      out[i] = data[i + 2];
      out[i + 1] = data[i + 1];
      out[i + 2] = data[i];
      out[i + 3] = data[i + 3];
    }
    return { width, height, rgba: out };
  }

  if (format === rawFormat.YUV420SP || format === rawFormat.YUV420P) {
    const need = width * height;
    if (data.length < need) return null;
    for (let i = 0; i < need; i += 1) {
      const y = data[i];
      const j = i * 4;
      out[j] = y;
      out[j + 1] = y;
      out[j + 2] = y;
      out[j + 3] = 255;
    }
    return { width, height, rgba: out };
  }

  return null;
}

export async function decodeJpegToRaw(frame) {
  if (!frame || frame.kind !== "jpg") throw new Error("decodeJpegToRaw requires a kind='jpg' image");
  if (typeof createImageBitmap !== "function" || typeof Blob === "undefined") {
    throw new Error("JPEG pixel decoding requires browser image APIs or a custom jpegDecoder");
  }
  const bitmap = await createImageBitmap(new Blob([toU8(frame.data)], { type: "image/jpeg" }));
  try {
    const width = Number(bitmap.width || 0);
    const height = Number(bitmap.height || 0);
    if (width <= 0 || height <= 0) throw new Error("decoded JPEG has invalid dimensions");
    const canvas = typeof OffscreenCanvas === "function"
      ? new OffscreenCanvas(width, height)
      : document.createElement("canvas");
    canvas.width = width;
    canvas.height = height;
    const context = canvas.getContext("2d", { willReadFrequently: true });
    if (!context) throw new Error("unable to create JPEG decode canvas");
    context.drawImage(bitmap, 0, 0);
    const pixels = context.getImageData(0, 0, width, height).data;
    return {
      kind: "raw",
      timestampNs: frame.timestampNs ?? 0n,
      width,
      height,
      format: RAW_FORMAT.RGBA32,
      channel: frame.channel || "preview",
      channelAlias: frame.channelAlias,
      data: new Uint8Array(pixels),
    };
  } finally {
    bitmap.close?.();
  }
}

export async function imageToRaw(frame, { jpegDecoder, includeReference = false } = {}) {
  if (!frame) return null;
  if (frame.kind === "raw") return frame;
  if (frame.kind === "stereo_raw") {
    const candidates = [frame.left, frame.right].filter(Boolean);
    const primary = candidates.find((candidate) => {
      const channel = String(candidate.channelAlias || candidate.channel || "").toLowerCase();
      return channel === "cam0" || channel === "preview" || channel === "left";
    });
    return primary || candidates[0] || null;
  }
  if (frame.kind !== "jpg") return null;
  if (frame.isReference && !includeReference) return null;
  const decoder = typeof jpegDecoder === "function" ? jpegDecoder : decodeJpegToRaw;
  return decoder(frame);
}

export async function decodeImageToRgb(frame, options = {}) {
  const raw = await imageToRaw(frame, options);
  return raw ? decodeRawToRgb(raw) : null;
}

File: python/mighty_sdk/image.py
from io import BytesIO
from typing import Any, Dict, Optional

import mighty_protocol as mp

from .utils import to_bytes


def select_primary_image(
    image: Optional[Dict[str, Any]],
    include_reference: bool = False,
) -> Optional[Dict[str, Any]]:
    """Return the mono/primary member of a Mighty image event."""
    if not image:
        return None
    kind = image.get("kind")
    if kind == "jpg" and image.get("is_reference") and not include_reference:
        return None
    if kind in ("raw", "jpg"):
        return image
    if kind != "stereo_raw":
        return None
    left = image.get("left") or {}
    right = image.get("right") or {}
    for candidate in (left, right):
        channel = str(
            candidate.get("channel_alias") or candidate.get("channel") or ""
        ).strip().lower()
        if channel in ("cam0", "preview", "left"):
            return candidate
    return left or right or None


def decode_jpeg_to_raw(
    image: Dict[str, Any],
    output_format: str = "gray8",
) -> Dict[str, Any]:
    """Decode a ``kind='jpg'`` event to a normal raw-image mapping.

    Pillow is imported lazily so receiving or forwarding compressed images does
    not add a mandatory image-library dependency to the transport SDK.
    """
    if not image or image.get("kind") != "jpg":
        raise ValueError("decode_jpeg_to_raw requires a kind='jpg' image")
    try:
        from PIL import Image
    except ImportError as exc:
        raise RuntimeError(
            "JPEG pixel decoding requires Pillow; install mighty-protocol[images]"
        ) from exc

    requested = str(output_format or "gray8").strip().lower()
    modes = {
        "gray8": ("L", mp.RAW_FORMAT["GRAY8"]),
        "rgb24": ("RGB", mp.RAW_FORMAT["RGB24"]),
        "rgba32": ("RGBA", mp.RAW_FORMAT["RGBA32"]),
    }
    if requested not in modes:
        raise ValueError("output_format must be gray8, rgb24, or rgba32")
    pillow_mode, raw_format = modes[requested]

    encoded = to_bytes(image.get("data", b""))
    if not encoded:
        raise ValueError("JPEG image data is empty")
    with Image.open(BytesIO(encoded)) as source:
        decoded = source.convert(pillow_mode)
        width, height = decoded.size
        data = decoded.tobytes()

    return {
        "kind": "raw",
        "timestamp_ns": int(image.get("timestamp_ns") or 0),
        "width": int(width),
        "height": int(height),
        "format": raw_format,
        "channel": str(image.get("channel") or "preview"),
        "channel_alias": image.get("channel_alias"),
        "data": data,
    }


def image_to_raw(
    image: Optional[Dict[str, Any]],
    jpeg_output_format: str = "gray8",
    include_reference: bool = False,
) -> Optional[Dict[str, Any]]:
    """Select the primary image and decode it when it is JPEG-compressed."""
    selected = select_primary_image(image, include_reference=include_reference)
    if not selected:
        return None
    if selected.get("kind") == "jpg":
        return decode_jpeg_to_raw(selected, jpeg_output_format)
    return selected
